runtime/lua/vim/treesitter/highlighter.lua

```diff
@@ -382,21 +380,35 @@ function TSHighlighter._on_spell_nav(_, _, buf, srow, _, erow, _)
 end
 
 ---@private
----@param _win integer
 ---@param buf integer
 ---@param topline integer
 ---@param botline integer
-function TSHighlighter._on_win(_, _win, buf, topline, botline)
+function TSHighlighter._on_win(_, _, buf, topline, botline)
   local self = TSHighlighter.active[buf]
-  if not self then
+  if not self or self.parsing then
+    return false
+  end
+  self.parsing = self.tree:parse({ topline, botline + 1 }, function(trees)
+    if trees then
+      self:_async_parse_callback(buf)
+    end
+  end) --[[@as boolean]]
+  if self.parsing then
```

Comment:
Does it work if `prepare_callback_states()` checks if individual trees are valid instead of using the flag?

Also IIUC, one unparsed injection currently makes the highlighter skip the window.